### PR TITLE
client, add validateClient to ClientBuilder

### PR DESCRIPTION
**Pull request by weidongxu-microsoft**
fix https://github.com/Azure/autorest.java/issues/2719

code change mainly in ServiceClientBuilderTemplate.java
https://github.com/Azure/autorest.java/pull/2783/files#diff-011f78c981243301193df6b93ff9aa9373f5486e722793f3ded47bd5425fcbf8

example of generated ClientBuilder
https://github.com/weidongxu-microsoft/autorest.java/blob/client_dpg-validate-builder/typespec-tests/src/main/java/com/resiliency/servicedriven/ResiliencyServiceDrivenClientBuilder.java#L258-L279

---

Would like advice on where we invoke the `validateBuilder()`.
Context: https://github.com/Azure/autorest.java/pull/2783#discussion_r1612764904
Basically, some variable like `credential` is on HttpPipeline, that should not be validated, if user provide their own HttpPipeline. And some variable like `endpoint` is not on HttpPipeline, that can always be validated when creating the client.
However, I don't want to make it more complex by providing a flag or 2 validate method (say `validatePipelineParameter`, `validateNonPipelineParameter`).
^^^
Updated to `validateClient` alone.
If dev need to do similar of `validatePipeline` for credential, they may do such
```java
if (this.pipeline == null) {
  // validate credential
}
```

### CODE REVIEW

File: javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java [hunk 1]
```diff
@@ -516,6 +529,20 @@ protected void addCreateHttpPipelineMethod(JavaSettings settings, JavaClass clas
...
+        addGeneratedAnnotation(classBlock);
+        classBlock.privateMethod("void validateBuilder()", methodBlock -> {
+            methodBlock.line("// This method is invoked from 'createHttpPipeline' when preparing the HTTP pipeline for the new client.");
+            methodBlock.line("// Developer can customize this method, to validate that the necessary conditions are met for the new client.");
+            for (ServiceClientProperty property : properties) {
+                // ServiceVersion is also provided with a defaultValueExpression
+                if (property.isRequired() && property.getDefaultValueExpression() == null) {
+                    methodBlock.line("Objects.requireNonNull(" + property.getName() + ", \"'" + property.getName() + "' cannot be null.\");");
+                }
+            }
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java:542]: We can choose not to automatically add the nonNull validation.

File: javagen/src/main/java/com/azure/autorest/template/TemplateHelper.java [hunk 2]
```diff
@@ -41,6 +41,7 @@ public static String getByteCloneExpression(String propertyName) {
     }
 
     public static void createHttpPipelineMethod(JavaSettings settings, String defaultCredentialScopes, SecurityInfo securityInfo, PipelinePolicyDetails pipelinePolicyDetails, JavaBlock function) {
+        function.line("this.validateBuilder();");
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/TemplateHelper.java:44]: `validateBuilder()` done in `createHttpPipeline()`.

One reason is that if user provide a `pipeline` instead, validation about "credential" does not need to happen (user could provide an HTTP pipeline that already integrated the credentials into pipeline.

However, the disadvantage here is that the validation on e.g. `endpoint` will also be skipped, if user provides the pipeline.

File: azure-tests/src/main/java/fixtures/azureparametergrouping/AutoRestParameterGroupingTestServiceBuilder.java [hunk 3]
```diff
@@ -241,8 +241,15 @@ public AutoRestParameterGroupingTestService buildClient() {
         return client;
     }
 
+    @Generated
+    private void validateBuilder() {
```
Comment (alzimmermsft) [inline azure-tests/src/main/java/fixtures/azureparametergrouping/AutoRestParameterGroupingTestServiceBuilder.java:245]: How often do we expect that developers will be customizing this method? I just wonder if it should be generated if there won't be any validations in it.

File: azure-tests/src/main/java/fixtures/custombaseuri/paging/AutoRestParameterizedHostTestPagingClientBuilder.java
Comment (alzimmermsft) [inline azure-tests/src/main/java/fixtures/custombaseuri/paging/AutoRestParameterizedHostTestPagingClientBuilder.java:1]: Not really attached to this file but something in all the builders.

We should replace

```java
HttpHeaders headers = new HttpHeaders();
        localClientOptions.getHeaders()
            .forEach(header -> headers.set(HttpHeaderName.fromString(header.getName()), header.getValue()));
        if (headers.getSize() > 0) {
            policies.add(new AddHeadersPolicy(headers));
        }
```

with `CoreUtils.createHttpHeadersFromClientOptions` as it has additional optimizations available due to it being within `azure-core` which can access some module specific things with headers.

File: javagen/src/main/java/com/azure/autorest/template/TemplateHelper.java (hunk 2 above)
Comment (alzimmermsft) [reply javagen/src/main/java/com/azure/autorest/template/TemplateHelper.java:44]: I wonder if we should have two methods for validation:

- `validateClient()` which validates anything used to create the client, but not in the pipeline. Ex, endpoint, service version, etc. Or, maybe this is generated into the client constructor.
- `validatePipeline()` which validates anything used to create the pipeline. Ex, credential information, etc.

File: azure-tests/src/main/java/fixtures/azureparametergrouping/AutoRestParameterGroupingTestServiceBuilder.java (hunk 3 above)
Comment (weidongxu-microsoft) [reply azure-tests/src/main/java/fixtures/azureparametergrouping/AutoRestParameterGroupingTestServiceBuilder.java:245]: From Bill, he mostly wants the validation on `endpoint` (which is usually "required" in ClientBuilder).

The other common case that I can think of, is for the credential. E.g. when SDK accept both KeyCredential and TokenCredential, client likely want to validate that one and only one of them is provided in ClientBuilder.

File: javagen/src/main/java/com/azure/autorest/template/TemplateHelper.java (hunk 2 above)
Comment (weidongxu-microsoft) [reply javagen/src/main/java/com/azure/autorest/template/TemplateHelper.java:44]: Emm, yes, that should work.
`validatePipeline()` would be called in `createHttpPipeline()`, `validateClient()` would be called in `createInnerClient()`.

@billwert Let us know your opinion as well.

File: azure-tests/src/main/java/fixtures/azureparametergrouping/AutoRestParameterGroupingTestServiceBuilder.java (hunk 3 above)
Comment (weidongxu-microsoft) [reply azure-tests/src/main/java/fixtures/azureparametergrouping/AutoRestParameterGroupingTestServiceBuilder.java:245]: https://github.com/Azure/autorest.java/pull/2785

File: azure-dataplane-tests/src/main/java/com/azure/ai/formrecognizer/documentanalysis/dpg/DocumentModelsClientBuilder.java [hunk 4]
```diff
@@ -262,8 +262,16 @@ private FormRecognizerClientImpl buildInnerClient() {
         return client;
     }
 
+    @Generated
+    private void validateBuilder() {
+        // This method is invoked from 'createHttpPipeline' when preparing the HTTP pipeline for the new client.
+        // Developer can customize this method, to validate that the necessary conditions are met for the new client.
```
Comment (haolingdong-msft) [inline azure-dataplane-tests/src/main/java/com/azure/ai/formrecognizer/documentanalysis/dpg/DocumentModelsClientBuilder.java:268]: Currently we will validate required fields not be null?

Comment (weidongxu-microsoft) [reply azure-dataplane-tests/src/main/java/com/azure/ai/formrecognizer/documentanalysis/dpg/DocumentModelsClientBuilder.java:268]: In current (main) code, we won't.

In this PR, see if we can add the default validation code, for required property without default value -- e.g. (typical case) the endpoint.

File: azure-tests/src/main/java/fixtures/custombaseuri/paging/AutoRestParameterizedHostTestPagingClientBuilder.java
Comment (weidongxu-microsoft) [reply azure-tests/src/main/java/fixtures/custombaseuri/paging/AutoRestParameterizedHostTestPagingClientBuilder.java:1]: done in a separate PR https://github.com/Azure/autorest.java/pull/2785

File: azure-dataplane-tests/src/main/java/com/azure/ai/formrecognizer/documentanalysis/dpg/DocumentModelsClientBuilder.java (hunk 4 above)
Comment (haolingdong-msft) [reply azure-dataplane-tests/src/main/java/com/azure/ai/formrecognizer/documentanalysis/dpg/DocumentModelsClientBuilder.java:268]: I mean this pr. Are we validating all the required property to be non-null, or only `endpoint` this time?

Comment (weidongxu-microsoft) [reply azure-dataplane-tests/src/main/java/com/azure/ai/formrecognizer/documentanalysis/dpg/DocumentModelsClientBuilder.java:268]: To all "required"

File: javagen/src/main/java/com/azure/autorest/template/TemplateHelper.java (hunk 2 above)
Comment (weidongxu-microsoft) [reply javagen/src/main/java/com/azure/autorest/template/TemplateHelper.java:44]: I would go `validateClient()` + `validatePipeline()`, if no further feedback.

Comment (weidongxu-microsoft) [reply javagen/src/main/java/com/azure/autorest/template/TemplateHelper.java]: Change to `validateClient()` alone, after discussion with Srikanta.

The thing in `validatePipeline()` maybe done as
```java
if (this.pipeline == null) {
  // validate credential
}
```
in `validateClient()`.

Comment (billwert) [reply javagen/src/main/java/com/azure/autorest/template/TemplateHelper.java]: yeah, I think just having `validateClient` is great. (Sorry I missed this ping!)

File: azure-tests/src/main/java/fixtures/azureparametergrouping/AutoRestParameterGroupingTestServiceBuilder.java (hunk 3 above)
Comment (billwert) [reply azure-tests/src/main/java/fixtures/azureparametergrouping/AutoRestParameterGroupingTestServiceBuilder.java]: I think generating it empty (if that happens) is fine; I assume the compiler is smart enough to elide empty function calls..

**Comment by haolingdong-msft** [comment]
`validateClient()` looks good to me. Just think of below:
1. wonder is there any reason to remove `validatePipeline()`
2. Later, we could collect the use case of validating clients, e.g. validate non-null fields. Provide examples in our docs to recommend user how to write good validation.

**Review by haolingdong-msft — APPROVED** [review]
